Integrate tag management to backend

ivov opened the pull request:
This PR adds endpoints for tag mangement:
- [x] GET `/tags`
- [x] POST `/tags` 
- [x] PATCH `/tags/:id`
- [x] DELETE `/tags/:id`

And it adjusts existing endpoints to accommodate tags:
- [x] GET `/workflows` now returns tags
- [x] GET `/workflows/:id` now returns tags
- [x] POST `/workflows` now accepts `tags=["1","2","3"]`
- [x] PATCH `/workflows/:id` now accepts `tags=["1","2","3"]`

Pending:
- [x] Remove operations with custom joining
- [x] Add operations with `relations`
- [x] Add table prefixes to query builder queries
- [x] Remove custom validators
- [x] Add validators from `class-validator`
- [x] Organize types
- [ ] Test query builder queries with 
  - [ ] Postgres
  - [ ] MySQL

Building on top of #1647

ivov committed:
:shirt: Appease linter

ivov committed:
:zap: Add tags retrieval endpoint

ivov committed:
:zap: Add tag creation endpoint

ivov committed:
:zap: Add tag deletion endpoint

ivov committed:
:zap: Add tag update endpoint

ivov committed:
:bug: Fix ID type in tag DB interface

ivov committed:
:hammer: Refactor param validation

ivov committed:
:zap: Add usage count to tags retrieval

ivov committed:
:shirt: Appease linter

ivov committed:
:zap: Add usage count to tags retrieval

ivov committed:
:zap: Add tagging and untagging endpoints

ivov committed:
:zap: Include tags in GET /workflows/:id

ivov committed:
:zap: Make usage count optional in GET /tags

ivov committed:
:zap: Add name validation to PATCH /tags/:id

ivov committed:
:zap: Add tags to /workflow

ivov committed:
:zap: Include tags in workflow creation response

ivov committed:
:zap: Add tags to PATCH /workflows/:id

ivov committed:
:hammer: Refactor ORM operations as helpers

ivov committed:
:hammer: Use helpers in POST /workflows

ivov committed:
:fire: Remove unused import

ivov committed:
:bug: Fix validators order

ivov committed:
:zap: Add tags to GET /workflows

ivov committed:
:hammer: Simplify types from helper return values

ivov committed:
:shirt: Fix linting issue

ivov committed:
:hammer: Refactor GET /tags for relations property

ivov committed:
:hammer: Simplify helpers with TypeORM features

ivov committed:
:hammer: Simplify workflow tags retrieval

ivov committed:
:fire: Remove unused imports

ivov committed:
Merge branch 'master' of https://github.com/n8n-io/n8n into add-tags-backend

ivov committed:
:zap: Stringify tag IDs in POST /workflows

ivov committed:
:zap: Stringify tag IDs in GET /workflows

ivov committed:
:zap: Stringify tag IDs in GET workflows/:id

ivov committed:
:zap: Stringify tag IDs in PATCH /workflows/:id

ivov committed:
:zap: Stringify tag IDs in GET /tags

ivov committed:
:zap: Stringify tag IDs in POST /tags

ivov committed:
Merge branch 'add-tags' of https://github.com/n8n-io/n8n into add-tags-backend

ivov committed:
:bug: Prevent zero-length tag names

mutdmour commented on packages/cli/src/Server.ts:
```diff
@@ -638,6 +663,25 @@ class App {
+			if (tags) {
+				await TagHelpers.deleteAllTagsForWorkflow(id);
+
+				const tagIds = tags.split(',');
+
+				for (const tagId of tagIds) {
+					await TagHelpers.validateId(tagId);
+				}
+
+				await TagHelpers.createRelations(id, tagIds);
+
+				const foundTags = await Db.collections.Tag!.find({
+					select: ['id', 'name'],
+					where: { id: In(tagIds) },
+				});
+				responseData.tags = foundTags.map(({ id, name }) => ({ id: id.toString(), name }));
```
can this be a utility function? since it's duplicated logic with adding

ivov replied:
They are not identical: the create version (as refactored now) does not call `validateNotRelated` and `removeRelations`. That said, the two versions can be consolidated in a single function with a flag for whether those two should be called. Is this advisable?

mutdmour replied:
I see. thanks. the current implementation you have is better.

mutdmour commented on packages/cli/src/Server.ts:
```diff
@@ -498,6 +500,18 @@ class App {
+
+			if (tags) {
+				const tagIds = tags.split(',');
+				await TagHelpers.createRelations(result.id as string, tagIds);
```
do we need to validate here first before creating relations like in updates?
https://github.com/n8n-io/n8n/pull/1648/files#diff-e1c41b1f8cc627242377b30fbee7af28472c81abe5f4f41da1c513f264f128e2R672

mutdmour commented on packages/cli/src/Server.ts:
```diff
@@ -711,6 +754,64 @@ class App {
+
+		// Creates a tag
+		this.app.post(`/${this.restEndpoint}/tags`, ResponseHelper.send(async (req: express.Request, res: express.Response): Promise<{ id: string, name: string }> => {
+			TagHelpers.validateRequestBody(req.body);
+
+			const { name } = req.body;
+			await TagHelpers.validateName(name);
+			TagHelpers.validateLength(name);
```
we can move all this validation under one function `validateName` since they all seem to be validating name property
```suggestion
			await TagHelpers.validateName(req.body);
```

mutdmour commented on packages/cli/src/Server.ts:
```diff
@@ -711,6 +754,64 @@ class App {
 		}));
 
+		// Retrieves all tags, with or without usage count
+		this.app.get(`/${this.restEndpoint}/tags`, ResponseHelper.send(async (req: express.Request, res: express.Response): Promise<ITagDb[] | Array<{ id: string, name: string, usageCount?: number }>> => {
```
just curious, how does authentication work here? is there a middleware somewhere that handles authentication for all rest endpoints? can you point me to that?

ivov replied:
Looks like it basic auth is implemented [here](https://github.com/n8n-io/n8n/blob/d59e6d1c4b41142495d5ca1ed0bdfdfcf6d66286/packages/cli/src/Server.ts#L224-L275) and JWT auth is implemented [here](https://github.com/n8n-io/n8n/blob/d59e6d1c4b41142495d5ca1ed0bdfdfcf6d66286/packages/cli/src/Server.ts#L277-L308). Docs reference [here](https://docs.n8n.io/reference/security.html).

mutdmour replied:
thanks for the info

mutdmour replied:
just wanted to confirm that authentication applies by default to these new endpoints. seems to be the case. 

only skipped for these endpoints https://github.com/n8n-io/n8n/blob/d59e6d1c4b41142495d5ca1ed0bdfdfcf6d66286/packages/cli/src/Server.ts#L219

mutdmour commented on packages/cli/src/TagHelpers.ts:
```diff
@@ -0,0 +1,165 @@
+/**
+ * Validate that a tag and a workflow are related so that their link can be deleted.
+ */
+export async function validateRelation(workflowId: number, tagId: number): Promise<void> | never {
```
this function does not seem to be used

mutdmour commented on packages/cli/src/TagHelpers.ts:
```diff
@@ -0,0 +1,165 @@
+/**
+ * Validate that a tag and a workflow are not related so that a link can be created.
+ */
+export async function validateNoRelation(workflowId: number, tagId: number): Promise<void> | never {
```
this function does not seem to be used

mutdmour commented on packages/cli/src/TagHelpers.ts:
```diff
@@ -0,0 +1,165 @@
+/**
+ * Check if a workflow and a tag are related. Used only in validators.
+ */
+async function checkRelated(
```
this can be removed as well

mutdmour reviewed (commented):
beautiful thanks for implementing this. just a few enhancements

ivov committed:
:hammer: Refactor per feedback

ivov commented on packages/cli/src/Server.ts:
```diff
@@ -498,6 +500,18 @@ class App {
 			// Save the workflow in DB
 			const result = await Db.collections.Workflow!.save(newWorkflowData);
 
+			const { tags } = req.body as { tags: string | undefined };
```
Implemented but extending `express.Request`.

ivov commented on packages/cli/src/Server.ts:
```diff
@@ -498,6 +500,18 @@ class App {
+					select: ['id', 'name'],
+					where: { id: In(tagIds) },
+				});
+				result.tags = foundTags.map(({ id, name }) => ({ id: id.toString(), name }));
```
Consolidated validators so that the whole request fails in that case.

mutdmour replied:
if I understand this correctly, based on current implementation, it would update the workflow first, then update the tags. and throw an error. I think you need to move validation above the updates

ivov replied:
Tags validation moved to the top in the latest iteration.

ivov commented on packages/cli/src/Server.ts:
```diff
@@ -498,6 +500,18 @@ class App {
+			const { tags } = req.body as { tags: string | undefined };
+
+			if (tags) {
+				const tagIds = tags.split(',');
```
Modified to accept `string[]`.

ivov commented:
@mutdmour Thanks for your feedback. Marked as resolved the points addressed by the refactoring.

mutdmour commented on packages/cli/src/TagHelpers.ts:
```diff
@@ -1,109 +1,128 @@
-	if (name.length <= 0 || name.length > 24) {
-		throw new ResponseHelper.ResponseError('Tag name must be 1 to 24 characters long.', undefined, 400);
-	}
+export function stringifyId(tags: ITagDb[]) {
```
nitpick - confusing naming.. might be a little clearer
```suggestion
export function getTagsResponse(tags: ITagDb[]) {
```

mutdmour commented on packages/cli/src/TagHelpers.ts:
```diff
@@ -1,109 +1,128 @@
-	if (!tag) {
-		throw new ResponseHelper.ResponseError(`Tag with ID ${id} does not exist.`, undefined, 400);
-	}
+function isStringArray(tags: unknown[]): tags is string[] {
+	return Array.isArray(tags) && !tags.some((value) => typeof value !== 'string');
```
nitpick - might be useful to move this to a utility file

mutdmour commented on packages/cli/src/TagHelpers.ts:
```diff
@@ -1,109 +1,128 @@
-// ----------------------------------
-//             queries
-// ----------------------------------
+	if (name.length <= 0 || name.length > 24) {
```
24 is magic number - so maybe add a variable to top of file or some other configuration file

```
const TAG_NAME_LENGTH_LIMIT = 24;
```

```suggestion
	if (name.length <= 0 || name.length > TAG_NAME_LENGTH_LIMIT) {
```

mutdmour reviewed (commented):
thanks for addressing my comments.. just a couple of minor issues

ivov committed:
:hammer: Move magic number to constant

ivov committed:
:art: Rename helper function

ivov committed:
:fire: Remove logging

ivov committed:
:hammer: Move tags validation to top

ivov committed:
:hammer: Refactor type guard for readability

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/n8n-io/n8n?pullRequest=1648) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/n8n-io/n8n?pullRequest=1648) before we can accept your contribution.<br/>**12** out of **13** committers have signed the CLA.<br/><br/>:white_check_mark: AllanDaemon<br/>:white_check_mark: ivov<br/>:white_check_mark: ahsanv<br/>:white_check_mark: cassvail<br/>:white_check_mark: theone74<br/>:white_check_mark: harshil1712<br/>:white_check_mark: janober<br/>:white_check_mark: RicardoE105<br/>:white_check_mark: MedAliMarz<br/>:white_check_mark: krynble<br/>:white_check_mark: bartvollebregt<br/>:white_check_mark: nosovk<br/>:x: snyk-bot<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/n8n-io/n8n?pullRequest=1648) it.</sub>

ivov committed:
:zap: Sort responses by request order

Sort the tags property in the responses of the POST /workflows and PATCH /workflows/:id by the order of the tag IDs in the request

ivov committed:
:zap: Fix error not thrown in relations validation

ivov committed:
Merge branch 'add-tags' into add-tags-backend

ivov committed:
:pencil2: Fix wording

ivov committed:
:fire: Remove unneeded variable

ivov committed:
:zap: Improve check for tags in POST request

ivov committed:
:hammer: Refactor async op in loop to Promise.all

ivov committed:
:zap: Remove validation covered by DB

ivov committed:
:hammer: Refactor POST /workflows to remove TypeORM-specific functionality

ivov committed:
:hammer: Overhaul to better incorporate TypeORM

ivov committed:
:zap: Refactor types to entities

ivov committed:
Merge branch 'add-tags' into add-tags-backend

ivov committed:
:zap: Ignore mistaken ID in POST /workflows

ivov committed:
:zap: Fix undefined tag ID in PATCH /workflows

ivov committed:
:zap: Shorten response for POST /tags

ivov committed:
:zap: Implement validation with class-validator

ivov committed:
:zap: Fix duplicate entry error for pg and MySQL

mutdmour commented on packages/cli/src/Server.ts:
```diff
@@ -484,34 +485,33 @@ class App {
+			newWorkflow.name = incomingData.name.trim();
 
-			const tagOrder = [...newWorkflowData.tags];
+			await this.externalHooks.run('workflow.create', [incomingData]);
```
might be useful to fire hook with tags

mutdmour commented on packages/cli/src/Server.ts:
```diff
@@ -484,34 +485,33 @@ class App {
 
-			await this.externalHooks.run('workflow.create', [newWorkflowData]);
+			Object.assign(newWorkflow, incomingData);
+			newWorkflow.name = incomingData.name.trim();
```
does name need to be validated before `trim` is applied?

ivov replied:
This is the workflow name. Validation logic in this PR is for tag names.

mutdmour commented on packages/cli/src/Server.ts:
```diff
@@ -711,6 +721,63 @@ class App {
+		}));
+
+		// Creates a tag
+		this.app.post(`/${this.restEndpoint}/tags`, ResponseHelper.send(async (req: express.Request, res: express.Response): Promise<IShortTag | void> => {
```
maybe we should add hooks to all these new tags endpoints similar to the workflow ones

BHesseldieck committed:
:hammer: refactor tag crud controllers

BHesseldieck committed:
🧹 remove unused imports

ivov committed:
:zap: Restore type to fix build

ivov committed:
:zap: Fix post-refactor PATCH /workflows/:id

ivov committed:
:zap: Fix PATCH /workflows/:id for zero tags

ivov committed:
:zap: Fix usage count becoming stringified

ivov commented:
Retested after refactoring of April 30

sqlite
- [x] GET `/tags`
- [x] POST `/tags`
- [x] PATCH `/tags/:id`
- [x] DELETE `/tags/:id`
- [x] GET `/workflows`
- [x] GET `/workflows/:id`
- [x] POST `/workflows`
- [x] PATCH `/workflows/:id` → fixed workflow/tag separation, fixed update to zero tags

MySQL
- [x] GET `/tags` → fixed `usageCount` becoming [stringified](https://github.com/typeorm/typeorm/issues/6196)
- [x] POST `/tags`
- [x] PATCH `/tags/:id`
- [x] DELETE `/tags/:id`
- [x] GET `/workflows`
- [x] GET `/workflows/:id`
- [x] POST `/workflows`
- [x] PATCH `/workflows/:id` → fixed workflow/tag separation, fixed update to zero tags

Postgres
- [x] GET `/tags` → fixed `usageCount` becoming [stringified](https://github.com/typeorm/typeorm/issues/6196)
- [x] POST `/tags`
- [x] PATCH `/tags/:id`
- [x] DELETE `/tags/:id`
- [x] GET `/workflows`
- [x] GET `/workflows/:id`
- [x] POST `/workflows`
- [x] PATCH `/workflows/:id` → fixed workflow/tag separation, fixed update to zero tags

BHesseldieck committed:
:zap: return tag id as string

BHesseldieck committed:
:hammer: add hooks to tag CUD operations

BHesseldieck committed:
🏎 simplify timestamp pruning

BHesseldieck committed:
Merge branch 'add-tags' into add-tags-backend

ivov committed:
:zap: Fix fs import to fix build
